src/darwin/Framework/CHIP/ServerEndpoint/MTRServerCluster.mm

```diff
@@ -0,0 +1,228 @@
+/**
+ *    Copyright (c) 2024 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#import "MTRDefines_Internal.h"
+#import "MTRDeviceController_Internal.h"
+#import "MTRLogging_Internal.h"
+#import "MTRServerAttribute_Internal.h"
+#import "MTRServerCluster_Internal.h"
+#import "MTRServerEndpoint_Internal.h"
+#import <Matter/MTRClusterConstants.h>
+#import <Matter/MTRServerCluster.h>
+
+#include <app/clusters/descriptor/descriptor.h>
+#include <lib/core/CHIPError.h>
+#include <lib/core/DataModelTypes.h>
+#include <lib/support/SafeInt.h>
+
+using namespace chip;
+
+MTR_DIRECT_MEMBERS
+@implementation MTRServerCluster {
+    /**
+     * The access grants our API consumer can manipulate directly.  Never
+     * touched on the Matter queue.
+     */
+    NSMutableSet<MTRAccessGrant *> * _accessGrants;
+    NSMutableArray<MTRServerAttribute *> * _attributes;
+    MTRDeviceController * __weak _deviceController;
```

Comment:
Is there no weak ref up to the MTRServerEndpoint? It seems like we could simply have the implementation mirror the logical structure, i.e. MTRServerAttribute -.cluster-> MTRServerCluster -.endpoint-> MTRServerEndpoint -.controller-> MTRDeviceController ?

These would simply be populated in the `add...:` method of the parent, and make the check for whether something already has a parent much more obvious.

Reply:
> Is there no weak ref up to the MTRServerEndpoint?

There isn't because I didn't really need it.  But also if we use a weak ref and use _that_ to detect added-or-not, that becomes non-deterministic (depends on when the parent happens to be deallocated).  That non-determinism can easily lead to "sometimes it works and sometimes it does not" code, and I don't think we want that.

So even if we had a weak backref I would need some other "keep track of whether I have been added" anyway.